mds/src/karatsuba_convolution.rs

```diff
@@ -0,0 +1,362 @@
+//! Calculate the convolution of two vectors using a Karatsuba-style
+//! decomposition and the CRT.
+//!
+//! Given a vector v \in F^N, let v(x) \in F[X] denote the polynomial
+//! v_0 + v_1 x + ... + v_{N - 1} x^{N - 1}.  Then w is equal to the
+//! convolution v * u if and only if w(x) = v(x)u(x) mod x^N - 1.
+//! Additionally, define the "signed convolution" by w(x) = v(x)u(x)
+//! mod x^N + 1.  Using the Chinese remainder theorem we can compute
+//! w(x) as
+//!     w(x) = 1/2 (w_0(x) + w_1(x)) + x^{N/2}/2 (w_0(x) - w_1(x))
+//! where
+//!     w_0 = v(x)u(x) mod x^{N/2} - 1
+//!     w_1 = v(x)u(x) mod x^{N/2} + 1
+//!
+//! To compute w_0 and w_1 we first compute
+//!                  v_0(x) = v(x) mod x^{N/2} - 1
+//!                  v_1(x) = v(x) mod x^{N/2} + 1
+//!                  u_0(x) = u(x) mod x^{N/2} - 1
+//!                  u_1(x) = u(x) mod x^{N/2} + 1
+//!
+//! Now w_0 is the convolution of v_0 and u_0 which we can compute
+//! recursively.  For w_1 we compute the signed convolution
+//! v_1(x)u_1(x) mod x^{N/2} + 1 using Karatsuba.
+//!
+//! There are 2 possible approaches to applying Karatsuba which mirror
+//! the DIT vs DIF approaches to FFT's.
+//!
+//! Option 1: left/right decomposition.
```

Comment:
Do we want to leave the discussion of option 1 here? We ended up going with option 2 so we could just discuss that one.